compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1502,6 +1506,145 @@ SmallVector<AffineMap> OnlineAttentionOp::getIndexingMapsArray() {
+                                   getMOffset());
+}
+
+void Im2colOp::setMixedKOffset(SmallVector<OpFoldResult> kOffset) {
+  SmallVector<int64_t> staticKOffset;
+  SmallVector<Value> dynamicKOffset;
+  dispatchIndexOpFoldResults(kOffset, dynamicKOffset, staticKOffset);
+  setStaticKOffset(staticKOffset);
+  getKOffsetMutable().assign(dynamicKOffset);
+}
+
+void Im2colOp::setMixedMOffset(SmallVector<OpFoldResult> mOffset) {
+  SmallVector<int64_t> staticMOffset;
+  SmallVector<Value> dynamicMOffset;
+  dispatchIndexOpFoldResults(mOffset, dynamicMOffset, staticMOffset);
+  setStaticMOffset(staticMOffset);
+  getMOffsetMutable().assign(dynamicMOffset);
+}
+
+/// Custom builder methods for im2col op.
+void Im2colOp::build(OpBuilder &builder, OperationState &state, Value input,
+                     Value output, ArrayRef<int64_t> strides,
+                     ArrayRef<int64_t> dilations,
+                     ArrayRef<OpFoldResult> kernelSize,
+                     ArrayRef<OpFoldResult> kOffset,
+                     ArrayRef<OpFoldResult> mOffset, ArrayRef<int64_t> batchPos,
+                     ArrayRef<int64_t> mPos, ArrayRef<int64_t> kPos) {
+  assert(strides.size() == kernelSize.size() &&
+         dilations.size() == kernelSize.size() &&
+         mPos.size() == kernelSize.size() &&
+         "strides, dilations, m_pos, and kernel expected to be the same rank");
+  SmallVector<int64_t> staticKernelSize, staticMOffset, staticKOffset;
+  SmallVector<Value> dynamicKernelSize, dynamicMOffset, dynamicKOffset;
+  dispatchIndexOpFoldResults(kernelSize, dynamicKernelSize, staticKernelSize);
+  dispatchIndexOpFoldResults(mOffset, dynamicMOffset, staticMOffset);
+  dispatchIndexOpFoldResults(kOffset, dynamicKOffset, staticKOffset);
+  SmallVector<Type> resultType;
+  auto outputType = output.getType();
+  if (isa<RankedTensorType>(outputType)) {
+    resultType.push_back(outputType);
+  }
+  build(builder, state, resultType, input, output,
+        builder.getDenseI64ArrayAttr(strides),
+        builder.getDenseI64ArrayAttr(dilations), dynamicKernelSize,
+        builder.getDenseI64ArrayAttr(staticKernelSize), dynamicKOffset,
+        builder.getDenseI64ArrayAttr(staticKOffset), dynamicMOffset,
+        builder.getDenseI64ArrayAttr(staticMOffset),
+        builder.getDenseI64ArrayAttr(batchPos),
+        builder.getDenseI64ArrayAttr(mPos), builder.getDenseI64ArrayAttr(kPos));
+}
+
+LogicalResult Im2colOp::verify() {
+  Operation *op = getOperation();
+  if (llvm::count_if(getDpsInputs(), [](Value v) {
+        return isa<ShapedType>(v.getType());
+      }) != 1) {
+    return op->emitOpError("expected only one ShapedType operand");
+  }
+  if (getNumDpsInits() != 1) {
+    return op->emitOpError("expected one output operand");
+  }
+  if (getMixedMOffset().size() != 1) {
+    return op->emitOpError("expected one m_offset");
+  }
+  if (getMixedKOffset().size() != 1) {
+    return op->emitOpError("expected one k_offset");
+  }
```

Comment:
It is a little weird that we allow multiple m_offset and k_offset in `.td` but we don't allow it in verifier. I understand that it can make the future extension a little easier (or maybe not?), but we should make it clear. Here is my suggestion:

1. If we want to allow variadic variables, we can add a TODO comment above to explain why it is not allowed now.
2. We only allow single m_offset and k_offset in `td` definition. I.e., use `Index` instead of `Variadic<Index>`.

I'm okay with both option, what do you think?

Reply:
Makes sense. Now that I consider it more, I think it is very likely that we will need multiple `k_offset` with data tiling (some sort of `NCHWc` layout with `FCHWc` filter), but it is unclear to me whether we will have multiple `m_offset`. Perhaps I will just switch to `Index` for now, and I will update it later when it is more clear.

Reply:
Actually, nevermind. I'm not sure how to get the right op assembly with plain `Index`. I'm just going to leave a TODO.